### PR TITLE
Use same git commit for all pipeline stages

### LINKED ISSUE #34598: Jenkins Pipeline Can Checkout Different Commits At Different Stages
**Describe the bug**

If a build of [Jenkins pipeline](https://builds.mantidproject.org/job/main_nightly_deployment_prototype/) is started and then part way through a pull request is merged to the `main` branch any further steps in the pipeline will pick up this new commit rather than building the same commit that started the pipeline.

**To Reproduce**

See for example https://builds.mantidproject.org/job/main_nightly_deployment_prototype/430/

The console log reports a failure to build `mantidworkbench` on Windows:

```sh
 File "C:\Jenkins\workspace\main_nightly_deployment_prototype\conda-recipes\recipes\mantidworkbench\meta.yaml", line 32, in top-level template code
     - setuptools {{ setuptools }}
   File "C:\Jenkins\workspace\main_nightly_deployment_prototype\mambaforge\envs\mantid-conda-build-env\lib\site-packages\conda_build\jinja_context.py", line 232, in pin_compatible
     pins, _, _ = get_env_dependencies(m, 'host', m.config.variant)
   File "C:\Jenkins\workspace\main_nightly_deployment_prototype\mambaforge\envs\mantid-conda-build-env\lib\site-packages\conda_build\render.py", line 138, in get_env_dependencies
     actions = environ.get_install_actions(tmpdir, tuple(dependencies), env,
   File "C:\Jenkins\workspace\main_nightly_deployment_prototype\mambaforge\envs\mantid-conda-build-env\lib\site-packages\boa\cli\mambabuild.py", line 133, in mamba_get_install_actions
     raise err
 conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform win-64: {MatchSpec("mantidqt=6.5.20221021.1050")}
 
 During handling of the above exception, another exception occurred:
```

**Expected behavior**

All build stages should use the same commit.

**Additional context**

https://issues.jenkins.io/browse/JENKINS-60591 notes a bug raised against Jenkins that matches this behaviour and suggests a workaround is to explicitly checkout the commit given by the `GIT_COMMIT` environment variable set by Jenkins.

### PR DESCRIPTION
**Pull request by thomashampson**
**Description of work.**
The main jenkins pipeline was picking up the latest git commit at each stage. If anything was pushed to main during the first conda packaging step (mantid, mantidqt, [and mantiddocs on linux only]), then the workbench packaging step would fail because the version numbers would be inconsistent.

The addition here will store the git commit at the start of the pipeline run and explicitly use this stored value to checkout the same commit at each stage. 

**To test:**
The following two builds had a commit pushed to the branch during the running of the first conda packaging stage. The first one was before the fix was implemented, the second was after. The first should have failed because it was looking for a mantid package that didn't exist, the second should succeed and build all linux packages.

https://builds.mantidproject.org/job/build_packages_from_branch/98/
https://builds.mantidproject.org/job/build_packages_from_branch/107/

Fixes #34598 

*This does not require release notes* because **it is not visibile to the end user**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile [hunk 1]
```diff
@@ -53,6 +53,18 @@ pipeline {
     WIN_BASH = "C:\\Program Files\\git\\bin\\bash.exe"
   }
   stages {
+    // Store the git commit hash in an environment variable so that we can use
+    // the same commit for all stages of the pipeline. This is perfomed by using
+    // the following step at the start of each subsequent stage:
+    //   checkout scm: [$class: 'GitSCM', branches: [[name: "${GIT_SHA}"]]]
```
Comment (martyngigg) [inline buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile:59]: Do we need to keep this commented out `checkout` command?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile (hunk 1 above)
Comment (thomashampson) [reply buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile:59]: I put it there to help explain why we are using that command in each of the other sections.

Comment (martyngigg) [reply buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile:59]: Oh yes I see now I have read the whole thing properly. Thanks.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could we get this merged so the pipeline build is protected when random merges to main happen. Thanks

**Pull request merged**